### ISSUE TITLE
Type1CConverter fails for PFB fonts

### ISSUE THREAD

**Issue by rbrito** [pts/pdfsizeopt#48]
Dear @pts,

I have one file here that I processed many years ago with Adobe's Acrobat 8 Pro and pdfsizeopt is having some difficulties to process it, because there is some problem with Ghostscript processing fonts:

```
(...)
info: loaded PDF of 2872506 bytes
info: using Ghostscript /usr/bin/gs: GPL Ghostscript 9.21 (2017-03-16)
info: decompressing 52 bytes with Ghostscript /Filter/FlateDecode/DecodeParms <</Columns 4/Predictor 12>>
info: decompressing 3802 bytes with Ghostscript /Filter/FlateDecode/DecodeParms <</Columns 5/Predictor 12>>
info: found 8395 obj offsets and 361 obj streams in xref stream
info: separated to 8032 objs + xref + trailer
info: parsed 8032 objs
info: found 3 Type1 fonts loaded
info: writing Type1CConverter (35528 font bytes) to: psotmp.22395.conv.tmp.ps
info: executing Type1CConverter with Ghostscript: gs -q -P- -dNOPAUSE -dBATCH -sDEVICE=pdfwrite -dPDFSETTINGS=/printer -dColorConversionStrategy=/LeaveColorUnchanged -sOutputFile=psotmp.22395.conv.tmp.pdf -f psotmp.22395.conv.tmp.ps
Type1CConverter: using interpreter GPL Ghostscript 921 20170316
Error: /syntaxerror in (bin obj seq, type=128, elements=1, size=59650, non-zero unused field)
Operand stack:
   --nostringval--   --nostringval--   --nostringval--
Execution stack:
   %interp_exit   .runexec2   --nostringval--   --nostringval--   --nostringval--   2   %stopped_push   --nostringval--   --nostringval--   --nostringval--   false   1   %stopped_push   1999   1   3   %oparray_pop   1998   1   3   %oparray_pop   1982   1   3   %oparray_pop   1868   1   3   %oparray_pop   --nostringval--   %errorexec_pop   .runexec2   --nostringval--   --nostringval--   --nostringval--   2   %stopped_push   --nostringval--   --nostringval--
Dictionary stack:
   --dict:1212/1684(ro)(G)--   --dict:0/20(G)--   --dict:99/200(L)--   --dict:0/9(L)--
Current allocation mode is local
Last OS error: No such file or directory
Current file position is 31090
GPL Ghostscript 9.21: Unrecoverable error, exit code 1
info: Type1CConverter failed, status=0x100
Traceback (most recent call last):
  File "/home/rbrito/Downloads/pdfsizeopt/pdfsizeopt", line 41, in <module>
    sys.exit(main.main(sys.argv, script_dir=script_dir))
  File "/home/rbrito/Downloads/pdfsizeopt/lib/pdfsizeopt/main.py", line 9092, in main
    pdf.ConvertType1FontsToType1C()
  File "/home/rbrito/Downloads/pdfsizeopt/lib/pdfsizeopt/main.py", line 5626, in ConvertType1FontsToType1C
    TMP_PREFIX + 'conv.tmp.ps', TMP_PREFIX + 'conv.tmp.pdf')
  File "/home/rbrito/Downloads/pdfsizeopt/lib/pdfsizeopt/main.py", line 5442, in GenerateType1CFontsFromType1
    assert False, 'Type1CConverter failed (status)'
AssertionError: Type1CConverter failed (status)
```

I'm using an up-to-date git clone. I am sending you a link to the problematic file privately.

Thanks in advance,

Rogério.

**Comment by pts** [comment]
Thank you for reprorting this!

It looks like Ghostscript is not able to load the font `/TheSansMono-Cd5`. However, when I process the PDF file with Ghostscript, it succeeds without font loading errors. Possible improvements to pdfsizeopt:

1. Make pdfsizeopt display a warning and continue, and leave the offending font unprocessed (as is).

2. Make pdfsizeopt call Ghostscript differently, with which Ghostscript is able to load the font.

I'll try to figure out if number 2 is possible.

**Comment by pts** [comment]
The offending PDF contains all 3 Type 1 fonts in PFB format rather than PFA format, and it makes Type1CConverter in pdfsizeopt fail. I think it's invalid PDF.

It is possible to make pdfsizeopt more resilient by making it accept PFB as well, because the conversion from PFB to PFA is very simple. I'm investigating the possible ways to do it.

FYI The error message `Error: /syntaxerror in (bin obj seq, type=128, elements=1, size=59650, non-zero unused field)` above indicates that Ghostscript is trying to run a PFB font program as PostScript code directly (without conversion to PFA first). There is no such error message for PFA font programs, because they are valid PostScript code.

**Comment by pts** [comment]
Fixed in 1986a7b6086f63bd4b5bfe9826dd72494d0297ab.

The fix was making Type1CConverter use fonts with `.loadfont` instead of `cvx exec`.
